pkg/actors/internal/membership.go

```diff
@@ -0,0 +1,283 @@
+	hostname   string
+
+	placementAddress   string
+	placementTables    *placement.ConsistentHashTables
+	placementTableLock *sync.RWMutex
+	placementConnAlive bool
+
+	placementSignal     chan struct{}
+	placementBlock      bool
+	placementClientCert *dapr_credentials.CertChain
+
+	operationUpdateLock *sync.Mutex
+
+	serviceHealthFn     func() bool
+	drainActorsFn       func()
+	evaluateRemindersFn func()
+}
+
+// NewActorMembership initializes ActorMembership for the actor service.
+func NewActorMembership(
+	placementAddress string, clientCert *dapr_credentials.CertChain,
+	appID, hostname string, actorTypes []string,
+	serviceHealthFn func() bool,
+	drainActorFn func(),
+	evaluateRemindersFn func()) *ActorMembership {
+	return &ActorMembership{
+		actorTypes:       actorTypes,
+		appID:            appID,
+		hostname:         hostname,
+		placementAddress: placementAddress,
+
+		placementTableLock:  &sync.RWMutex{},
+		placementTables:     &placement.ConsistentHashTables{Entries: make(map[string]*placement.Consistent)},
+		placementClientCert: clientCert,
+		operationUpdateLock: &sync.Mutex{},
+
+		serviceHealthFn:     serviceHealthFn,
+		drainActorsFn:       drainActorFn,
+		evaluateRemindersFn: evaluateRemindersFn,
+	}
+}
+
+// Start connects placement service to register to membership and send heartbeat
+// to report the current member status periodically.
+func (am *ActorMembership) Start() {
+	// placementConnAlive represents the status of stream channel. This must be changed in receiver loop.
+	// This flag reduces the unnecessary request retry.
+	am.placementConnAlive = true
+	stream := am.newPlacementStreamConn(am.placementAddress)
+
+	// Establish receive channel to retrieve placement table update
+	go func() {
+		for {
+			// Wait until stream is reconnected.
+			if !am.placementConnAlive || stream == nil {
+				time.Sleep(placementReconnectInterval)
+				continue
+			}
+
+			resp, err := stream.Recv()
+			if err != nil {
+				diag.DefaultMonitoring.ActorStatusReportFailed("recv", "status")
+				log.Warnf("failed to receive placement table update from placement service: %v", err)
+
+				// If receive channel is down, ensure the current stream is closed and get new gRPC stream.
+				stream.CloseSend()
+				am.placementConnAlive = false
+				stream = am.newPlacementStreamConn(am.placementAddress)
+				am.placementConnAlive = true
+
+				continue
+			}
+
+			diag.DefaultMonitoring.ActorStatusReported("recv")
+			am.onPlacementOrder(resp)
+		}
+	}()
+
+	// Send the current host status to placement to register the member and
+	// maintain the status of member by placement.
+	go func() {
+		for {
+			// Wait until stream is reconnected.
+			if !am.placementConnAlive || stream == nil {
+				time.Sleep(placementReconnectInterval)
+				continue
+			}
+
+			// serviceHealthFn() is the health status of actor service application. This allows placement to update
+			// memberlist and hashing table quickly.
+			if !am.serviceHealthFn() {
```

Comment:
is there a possibility of hitting this function multiple times in a second when the actor service is unhealthy? Since the loop is continued without a sleep?